torch/_prims_common/__init__.py

```diff
@@ -454,10 +454,12 @@ def canonicalize_dim(rank: int, idx: int) -> int:
     else:
         _idx = idx
 
-    if _idx < 0 or _idx > _rank:
+    if _idx < 0 or _idx >= _rank:
+        # Always report the lower dimension first
+        bound1, bound2 = sorted((-rank, rank - 1))
```

Comment:
why do you need that? Rank is always 1 or more, so `-rank, rank-1` is always sorted

Reply:
good catch! i should have looked more closely into why that even happens.

without `sorted` (note the bounds order here):

```py
In [1]: import torch

In [2]: torch.diagonal(torch.tensor(0))
[...]
IndexError: Dimension out of range (expected to be in range of [-1, 0], but got 1)

In [3]: torch._refs.diagonal(torch.tensor(0))
[...]
IndexError: Dimension out of range (expected to be in range of [0, -1], but got 1)
```

this happens because `rank` is `0` in this case, so `(-rank, rank - 1)` evaluates to `(0, -1)`:
```
In [4]: torch.tensor(0).ndim
Out[4]: 0
```

**the real issue** (as has been pointed out by you somewhere in the stack earlier) is that `canonicalize_dim` doesn't support scalars.

the line here:

```py
_rank = rank if rank != 0 else 1
```

is supposed to be this, i think:

```cpp
int64_t maybe_wrap_dim_slow(
    int64_t dim,
    int64_t dim_post_expr,
    bool wrap_scalar) {
  if (dim_post_expr <= 0) {
    TORCH_CHECK_INDEX(
        wrap_scalar,
        "dimension specified as ",
        dim,
        " but tensor has no dimensions");
    return c10::maybe_wrap_dim(dim, /*dim_post_expr=*/1, /*wrap_scalar=*/false);
```

so i'm going to push this change now, which tries to mimic this behavior more closely, giving us more flexibility there:

```py
def canonicalize_dim(rank: int, idx: int, wrap_scalar: bool = True) -> int:
    if rank <= 0:
        if not wrap_scalar:
            msg = f"dimension specified as {idx} but tensor has no dimensions"
            raise IndexError(msg)
        rank = 1 

    if idx >= 0 and idx < rank:
        return idx

    if idx < 0:
        _idx = idx + rank
    else:
        _idx = idx

    if _idx < 0 or _idx >= rank:
        # Same error message as in aten/src/ATen/WrapDimUtils.h:49
        msg = "Dimension out of range (expected to be in range of [{0}, {1}], but got {2})".format(
            -rank, rank -1, idx
        )
        raise IndexError(msg)

    return _idx
```

with the above definition, this test passes (it's based on my earlier port of the c++ functions to python), as well as the testsuite for functions in this stack:

<details>
  <summary>Test code here</summary>

```py
import torch
from itertools import product

from typing import Callable, Type

def check(
    b: bool, s: Callable[[], str], exc_type: Type[Exception] = RuntimeError
) -> None:
    """
    Helper function for raising an error_type (default: RuntimeError) if a boolean condition fails.
    Error message is a callable producing a string (to avoid wasting time
    string formatting in non-error case, and also to make it easier for torchdynamo
    to trace.)
    """
    if not b:
        raise exc_type(s())

# See maybe_wrap_dim in core
def _maybe_wrap_dim(
    idx: int,
    rank: int,
    wrap_scalar: bool = True,
) -> int:
    # if dim_post_expr is 0 and wrap_scalar is true, then dim must be in the
    # range [-1, 0]. This is a special case for scalar tensors and manifests in
    # e.g. torch.sum(scalar_tensor, 0) Otherwise, dim should be in the range
    # [-dim_post_expr, dim_post_expr-1].
    dim = idx
    dim_post_expr = rank

    # Fast path
    if -dim_post_expr <= dim and dim < dim_post_expr:
        # Branch-less version of dim + (dim < 0 ? dim_post_expr : 0)
        return dim + dim_post_expr * (dim < 0)

    # Check edge-cases out-of-line (wrapping scalars and out-of-bounds errors)
    return _maybe_wrap_dim_slow(dim, dim_post_expr, wrap_scalar)


def _maybe_wrap_dim_slow(
    dim: int,
    dim_post_expr: int,
    wrap_scalar: bool,
) -> int:
    if dim_post_expr <= 0:
        check(
            wrap_scalar,
            lambda: f"dimension specified as {dim} but tensor has no dimensions",
        )
        idx = dim
        return _maybe_wrap_dim(idx=idx, rank=1, wrap_scalar=False)

    min = -dim_post_expr
    max = dim_post_expr - 1

    check(
        min <= dim and dim <= max,
        lambda: (
            "Dimension out of range (expected to be in range of "
            f"[{min}, {max}], but got {dim})"
        ),
        exc_type=IndexError,
    )

    raise RuntimeError("should never reach here as dim should be out-of-bounds")


wrap_scalars = (True, False)
ranks = range(-3, 6)
idxs = range(-3, 8)


def test(func, rank, idx, wrap_scalar):
    try:
        res = func(rank=rank, idx=idx, wrap_scalar=wrap_scalar)
    except Exception as e:
        res = str(e)

    return res


def run_tests():
    for rank, idx, wrap_scalar in product(ranks, idxs, wrap_scalars):
        res1 = test(torch._prims_common.canonicalize_dim, rank=rank, idx=idx, wrap_scalar=wrap_scalar)
        res2 = test(_maybe_wrap_dim, idx=idx, rank=rank, wrap_scalar=wrap_scalar)
        print("RES1", res1)
        print("RES2", res2)
        if res1 != res2:
            print("!!! ERROR !!!")
            print(res1)
            print(res2)
            return
    print("!!! ALL DONE !!!")


run_tests()
```

</details>